src/sentry/snuba/metrics.py

```diff
@@ -68,41 +68,76 @@
 TS_COL_GROUP = "bucketed_time"
 
 
+class MetricIndexNotFound(InvalidParams):
+    pass
+
+
 def reverse_resolve(index: int) -> str:
+    assert index != 0
     resolved = indexer.reverse_resolve(index)
     # The indexer should never return None for integers > 0:
-    assert resolved is not None
+    if resolved is None:
+        raise MetricIndexNotFound()
 
     return resolved
 
 
-def reverse_resolve_groupby(index: int) -> Optional[str]:
+def reverse_resolve_weak(index: int) -> Optional[str]:
+    """
+    Resolve an index value back to a string, special-casing 0 to return None.
+
+    This is useful in situations where a `GROUP BY tags[123]` clause produces a
+    tuple for metric buckets that are missing that tag, i.e. `tags[123] == 0`.
+    """
+
     if index == 0:
-        # When a groupBy is requested with a tag that does not exist for the given
-        # metric, tags[i] == 0. In this case, return None
         return None
 
     return reverse_resolve(index)
 
 
-def resolve_tag_key(string: str) -> str:
+def resolve(string: str) -> int:
     resolved = indexer.resolve(string)
     if resolved is None:
-        raise InvalidParams(f"Unknown tag key: '{string}'")
+        raise MetricIndexNotFound(f"Unknown tag key: '{string}'")
```

Comment:
The exception message for the general case should probably not be "Unknown tag key", but for responding to the API client it is useful information. Maybe `resolve_tag_key` should catch `MetricIndexNotFound` and convert it to `InvalidParams` with a specific message (instead of deriving one exception from the other)?

Reply:
I would still like resolve_tag_key to raise the same exception type as other methods, but yeah this exception message does not make a ton of sense.

How useful is this information really, can we get away with just changing this to "Unknown string"?

Reply:
I can live with that.